### PR TITLE
Quadhoral openmeteo alex

### LINKED ISSUE #20: Test Quadhoral OpenMeteo Extract.py
As a developer, I want to test the Extract.py script to ensure that weather data is reliably retrieved from the API, so I can trust that the pipeline captures accurate and complete information necessary for stargazing predictions.

- Using `pytest`

### PR DESCRIPTION
**Pull request by AlexOBarnes**
This closes #20 #13 

What I have done:
- Created `test_quadhoral_pipeline.py`
- Written tests for `quadhoral_pipeline.py`
- Implemented functions to pass tests
- Created `conftest.py`

### CODE REVIEW

File: quadhoral-openmeteo/api_error.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+'''This module contains the API error class'''
+
+class APIError(Exception):
```
Comment (MurasatoNaoya) [inline quadhoral-openmeteo/api_error.py:3]: Similar to before, good use of error classes to categorise errors in an OOP fashion.

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 2]
```diff
@@ -0,0 +1,128 @@
+'''Contains tests for the extract script in this folder'''
+import pytest
+from unittest.mock import patch, MagicMock
+from quadhoral_extract import get_county_coordinates, convert_to_params, request_weather_data, extract
+from api_error import APIError
+
+class TestGetCountyCoordinates:
+
+    @patch('quadhoral_extract.get_connection')
```
Comment (MurasatoNaoya) [inline quadhoral-openmeteo/test_quadhoral_extract.py:9]: Good use of patching for idempotency, avoiding actually interacting with the OpenMeteoAPI or writing to our RDS.

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 3]
```diff
@@ -0,0 +1,128 @@
+'''Contains tests for the extract script in this folder'''
```
Comment (MurasatoNaoya) [inline quadhoral-openmeteo/test_quadhoral_extract.py:1]: Good use of test classes, to categorise tests into related groups. Also very readable this way.

**Review by MurasatoNaoya — APPROVED** [review]
LGTM. A lot of these pipelines have very similar functionality, but there are still many aspects to fetching weather data specifically that this pipeline does very well.

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        with pytest.raises(ValueError, match="Missing Values in returned data"):
+            extract()
+
+    @patch('quadhoral_extract.get_county_coordinates')
+    def test_extract_db_error(self, mock_coords):
+        """Test extraction pipeline fails due to DB connection."""
+        mock_coords.side_effect = Exception("Database connection error")
+
+        with pytest.raises(Exception, match="Database connection error"):
+            extract()
```
Comment (AlexOBarnes) [inline quadhoral-openmeteo/test_quadhoral_extract.py:128]:
```suggestion
class TestExtract:

    @patch('quadhoral_extract.get_county_coordinates')
    @patch('quadhoral_extract.convert_to_params')
    @patch('quadhoral_extract.request_weather_data')
    @patch.dict(os.environ, {"DB_NAME": "test_db", "DB_USER": "test_user", "DB_PASS": "test_pass"})
    def test_extract_valid(self, mock_weather_data, mock_params, 
                                           mock_coords, valid_json_data):
        """Test successful data extraction pipeline."""
        mock_coords.return_value = ([1.0], [51.0])
        mock_params.return_value = {'latitude': [1.0], 'longitude': [51.0]}
        mock_weather_data.return_value = valid_json_data

        result = extract()
        assert len(result) == 106

    @patch('quadhoral_extract.request_weather_data')
    @patch.dict(os.environ, {"DB_NAME": "test_db", "DB_USER": "test_user", "DB_PASS": "test_pass"})
    def test_extract_no_data(self, mock_weather_data):
        """Test no data returned in extraction pipeline."""
        mock_weather_data.side_effect = ValueError(
            "Missing Values in returned data")

        with pytest.raises(ValueError, match="Missing Values in returned data"):
            extract()

    @patch('quadhoral_extract.get_county_coordinates')
    @patch.dict(os.environ, {"DB_NAME": "test_db", "DB_USER": "test_user", "DB_PASS": "test_pass"})
    def test_extract_db_error(self, mock_coords):
        """Test extraction pipeline fails due to DB connection."""
        mock_coords.side_effect = Exception("Database connection error")

        with pytest.raises(Exception, match="Database connection error"):
            extract()
```

**Review by AlexOBarnes — COMMENTED** [review]
This needs a change to pass pytest action

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 5]
```diff
@@ -0,0 +1,132 @@
+'''Contains tests for the extract script in this folder'''
+import pytest
+from unittest.mock import patch, MagicMock
```
Comment (AlexOBarnes) [inline quadhoral-openmeteo/test_quadhoral_extract.py:3]:
```suggestion
'''Contains tests for the extract script in this folder'''
import os
import pytest
from unittest.mock import patch, MagicMock
```

**Review by AlexOBarnes — COMMENTED** [review]
Additional changes needed

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+    @patch('quadhoral_extract.request_weather_data')
+    @patch.dict(os.environ, {"DB_NAME": "test_db", "DB_USER": "test_user", "DB_PASS": "test_pass"})
+    def test_extract_no_data(self, mock_weather_data):
+        """Test no data returned in extraction pipeline."""
+        mock_weather_data.side_effect = ValueError(
+            "Missing Values in returned data")
+
+        with pytest.raises(ValueError, match="Missing Values in returned data"):
+            extract()
```
Comment (AlexOBarnes) [inline quadhoral-openmeteo/test_quadhoral_extract.py:124]:
```suggestion
    @patch.dict(os.environ, {
        "DB_NAME": "test_db",
        "DB_USER": "test_user",
        "DB_PASS": "test_pass",
        "DB_HOST": "localhost",
        "DB_PORT": "5432"
    })
    @patch('quadhoral_extract.request_weather_data')
    def test_extract_no_data(self, mock_weather_data):
        """Test no data returned in extraction pipeline."""
        mock_weather_data.side_effect = ValueError("Missing Values in returned data")

        with pytest.raises(ValueError, match="Missing Values in returned data"):
            extract()
```

**Review by AlexOBarnes — COMMENTED** [review]
Final change hopefully

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        mock_params.return_value = {'latitude': [1.0], 'longitude': [51.0]}
+        mock_weather_data.return_value = valid_json_data
+
+        result = extract()
+        assert len(result) == 106
+
+    @patch.dict(os.environ, {
+        "DB_NAME": "test_db",
+        "DB_USER": "test_user",
+        "DB_PASS": "test_pass",
```
Comment (AlexOBarnes) [inline quadhoral-openmeteo/test_quadhoral_extract.py:119]:
```suggestion
        "DB_PASSWORD": "test_pass",
```

**Review by AlexOBarnes — COMMENTED** [review]
This is spelled wrong

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        "DB_HOST": "localhost",
+        "DB_PORT": "5432"
+    })
+    @patch('quadhoral_extract.request_weather_data')
+    def test_extract_no_data(self, mock_weather_data):
+        """Test no data returned in extraction pipeline."""
+        mock_weather_data.side_effect = ValueError("Missing Values in returned data")
+
+        with pytest.raises(ValueError, match="Missing Values in returned data"):
+            extract()
```
Comment (AlexOBarnes) [inline quadhoral-openmeteo/test_quadhoral_extract.py:129]:
```suggestion
 
```

**Review by AlexOBarnes — COMMENTED** [review]
Final change

File: quadhoral-openmeteo/api_error.py (hunk 1 above)
Comment (olliejth) [reply quadhoral-openmeteo/api_error.py:3]: I concur.

File: quadhoral-openmeteo/api_error.py [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+            401: "Unauthorized: Authentication is required and has failed.",
+            403: "Forbidden: You do not have permission to access this resource.",
+            404: "Not Found: The requested resource could not be found.",
+            405: "Method Not Allowed: The request method is not supported for this resource.",
+            408: "Request Timeout: The server timed out waiting for the request.",
+            429: "Too Many Requests: You have sent too many requests in a given amount of time.",
+            500: "Internal Server Error: The server encountered an unexpected condition.",
+            502: "Bad Gateway: The server received an invalid response from the upstream server.",
+            503: "Service Unavailable: The server is currently unable to handle the request.",
+            504: "Gateway Timeout: The server did not respond in a timely manner."}
```
Comment (CalebMathers) [inline quadhoral-openmeteo/api_error.py:36]: Good coverage of error codes

File: quadhoral-openmeteo/test_quadhoral_extract.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        with pytest.raises(ValueError, match="Missing Values in returned data"):
+            request_weather_data({})
+
+
+class TestExtract:
+
+    @patch('quadhoral_extract.get_county_coordinates')
+    @patch('quadhoral_extract.convert_to_params')
+    @patch('quadhoral_extract.request_weather_data')
+    @patch.dict(os.environ, {"DB_NAME": "test_db", "DB_USER": "test_user", "DB_PASS": "test_pass"})
```
Comment (CalebMathers) [inline quadhoral-openmeteo/test_quadhoral_extract.py:105]: good patching!

**Review by CalebMathers — APPROVED** [review]

**Pull request merged**